base/compiler/abstractinterpretation.jl

```diff
@@ -2557,8 +2642,13 @@ function abstract_call(interp::AbstractInterpreter, arginfo::ArgInfo, sv::Infere
         add_curr_ssaflag!(sv, IR_FLAG_UNUSED)
     end
     si = StmtInfo(!unused)
-    call = abstract_call(interp, arginfo, si, sv)
-    sv.stmt_info[sv.currpc] = call.info
+    call = abstract_call(interp, arginfo, si, sv)::Future
+    Future{Nothing}(call.completed, interp, sv) do interp, sv
+        # this only is needed for the side-effect, sequenced before any task tries to consume the return value,
+        # which this will do even without returning this Future
+        sv.stmt_info[sv.currpc] = call[].info
+        nothing
+    end
```

Comment:
Dealing with this kind of complexity is really tricky, but it might be something we can't avoid as we move forward with this design.

Reply:
The comment is not quite as tricky as it sounds anymore, since the ordering is enforced by presence in the workqueue, not the consume for the return value